### PR TITLE
skip endpointUpdate if endpoint checksum stays the same

### LINKED ISSUE #26637: kube-proxy should ignore "spam" updates to endpoints
#### Environment
- 3 node Kubernetes v1.2.4 cluster with HA
- scheduler and controller-manager have `--leader-elect=true`
- no additional POD running aside from those in the `kube-system` namespace
#### Observed behavior
- `kube-proxy` runs `iptables-save/restore` twice per second.

Increasing `kube-proxy` verbosity shows that this is triggered by `OnEndpointsUpdate`

Stopping all the schedulers and controller-managers, then restarting only one instance of each without `--leader-elect` stops `kube-proxy` from running `iptables` that often.

### PR DESCRIPTION
**Pull request by freehan**
fixes: https://github.com/kubernetes/kubernetes/issues/26637

### CODE REVIEW

**Comment by freehan** [comment]
Ready for review. 

OnEndpointsUpdate is responsive to endpoint changes. Proxy iptables rules are forced syncd every 30 seconds.

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -588,6 +591,11 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
 	defer proxier.mu.Unlock()
 	proxier.haveReceivedEndpointsUpdate = true
 
+	if !proxier.needToUpdate(allEndpoints) {
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: don't you also want to call this from OnServiceUpdate?

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
+	checksumMap := make(map[string]uint32)
+	hash := fnv.New32a()
+
+	for _, ep := range allEndpoints {
+		hashutil.DeepHashObject(hash, ep.Subsets)
+		checksumMap[fmt.Sprintf("%s/%s", ep.Namespace, ep.Name)] = hash.Sum32()
+	}
+
+	needToUpdate := false
+	if len(checksumMap) == len(proxier.endpointChecksums) {
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: nit: I'd prefer an early reaturn instead
```go
if len(checksum) != len(proxier.endpoint) {
  return true
}
```

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
+// needToUpdate returns if the input endpoints is different from the last batch
+// This is calculated by comparing the checksum of EndpointSubsets.
+func (proxier *Proxier) needToUpdate(allEndpoints []api.Endpoints) bool {
+	checksumMap := make(map[string]uint32)
+	hash := fnv.New32a()
+
+	for _, ep := range allEndpoints {
+		hashutil.DeepHashObject(hash, ep.Subsets)
+		checksumMap[fmt.Sprintf("%s/%s", ep.Namespace, ep.Name)] = hash.Sum32()
+	}
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: Might be simpler to defer proxier.endpointChecksums = checksumMap here

File: pkg/proxy/iptables/proxier.go [hunk 4]
```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
+				needToUpdate = true
+				break
+			}
+		}
+	} else {
+		needToUpdate = true
+	}
+
+	if needToUpdate {
+		proxier.endpointChecksums = checksumMap
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: I guess this method is not thread safe. If you call it from both OnServiceUpdate and OnEndpointUpdate, can they race? or are they always invoked serially? If there's a possibility of a race you should wrap this in setEndpointChecksums, getEndpointChecksums which does the expected set/get under a lock

Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: One way to refactor thread safety and get/set is (doesn't matter if thread safety is not an issue)
```go
type endpointHasher struct {
  oldEndpoints 
}

// setEndpoints sets the internal field
// not threadsafe
func (e *endpointHasher) setEndpoints(newEndpoints) {}

// isEqual compares a hash of the internal field with the hash of the given newEndpoints
// not threadsafe
func (e *endpointHasher) isEqual(newEndpoints) {}

// ShouldUpdate returns true if the newEndpoints require an update.
// Also stores newEndpoints in the internal field.
func (e *endpointHasher) ShouldUpdate(newEndpoints) {
   e.lock.Lock()
   defer e.lock.Unlock()

   if e.isEqual() {
    return true
  }

  e.setEndpoints(newEndpoints)
  return false
}
```

**Review by bprashanth — COMMENTED** [review]
Main concern is thread safety and not invoking the hash function from onServiceUpdate

File: pkg/proxy/iptables/proxier.go [hunk 5]
```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
 	proxier.deleteEndpointConnections(staleConnections)
 }
 
+// needToUpdate returns if the input endpoints is different from the last batch
+// This is calculated by comparing the checksum of EndpointSubsets.
+func (proxier *Proxier) needToUpdate(allEndpoints []api.Endpoints) bool {
+	checksumMap := make(map[string]uint32)
+	hash := fnv.New32a()
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: I'm no expert on the fnv family of hash functions but the implementation seems to mutate the value such that one probably needs to invoke reset: https://golang.org/src/hash/fnv/fnv.go (though I'm assuming you verified it, just checking)

**Comment by freehan** [comment]
closing this one in favor of #41223

**Pull request closed without merging**